Error: unable to find valid certification path to requested target

https://github.com/JetBrains/azure-tools-for-intellij/discussions/777#discussioncomment-8381402

Hi,

I have exactly the same problem, but it doesn't appear to be related to Azurite. 

I've checked the Azurite settings, and there's no problem spinning it up or connecting to it. I even took the step of creating a certificate using the `mkcert` command, and also removing Azurite from the 'pre launch' config, but still no dice.

I did some more digging and came across a youtrack article on a search engine which seemed to be exactly what I was experiencing, but the link went to an 'issue not found'. However the description (see screenshot of summary below) pretty much described exactly what I was experiencing.  

![image](https://github.com/user-attachments/assets/3ec61dea-9c3b-4d8d-9897-231dd0148e61)

This is the link it was pointing to: https://youtrack.jetbrains.com/issue/RIDER-117246

I've combed through YouTrack trying to find it, perhaps it's been deleted.

It was working previously, but seemed to break around the upgrade to Rider 2024.2.4 (I have subsequently updated to 2024.2.5 but the problem persists). I also have the latest version of the Azure Toolkit for Rider plugin (v4.2.0), and have tried downgrading that plugin too, also with no luck.

The same function works perfectly (and always has done) under Visual Studio 2022. It's easily reproducible for me:

* Create a new Azure function, with any trigger (I used http), using either VS or Rider. It doesn't matter.
* Build it.
* Run it. Visual Studio works, Rider does not (but used to).

The error, when it comes up if I hit Run or Debug (it doesn't matter) is this:

![image](https://github.com/user-attachments/assets/4aff2743-c2c8-4fac-9573-59aa5588bcbe)

But that is literally ALL it gives you, there is _no_ indication in the error as to what has happened.

I found the error in the idea.x.log files, and it seems to be related to the Azure Toolkit plugin, but I am not sure _why_ (starts at the `SEVERE` entry, below):

```text
2024-09-21 13:04:59,215 [  18008]   INFO - #c.i.p.w.WhatsNewShowOnStartCheckService - Got What's New content: com.intellij.platform.whatsNew.WhatsNewVisionContent@11f2446d
2024-09-21 13:04:59,224 [  18017]   INFO - #c.i.p.w.WhatsNewContentVersionChecker - Comparing versions 2024-2.5-22855-7890de2744a5f60a239878afaffeaa01425f5c54 and 2024-2.3-21829-7890de2744a5f60a239878afaffeaa01425f5c54: false.
2024-09-21 13:04:59,224 [  18017]   INFO - #c.i.p.w.WhatsNewShowOnStartCheckService - Should show What's New: false
2024-09-21 13:04:59,278 [  18071] SEVERE - #c.i.i.s.i.StartupManagerImpl - unable to find valid certification path to requested target [Plugin: com.intellij.resharper.azure]
com.intellij.diagnostic.PluginException: unable to find valid certification path to requested target [Plugin: com.intellij.resharper.azure]
	at com.intellij.ide.startup.impl.StartupManagerImplKt$launchActivity$1.invokeSuspend(StartupManagerImpl.kt:502)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.UndispatchedCoroutine.afterResume(CoroutineContext.kt:277)
	at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:99)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:102)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:608)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:873)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:763)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:750)
Caused by: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at java.base/sun.security.provider.certpath.SunCertPathBuilder.build(SunCertPathBuilder.java:148)
	at java.base/sun.security.provider.certpath.SunCertPathBuilder.engineBuild(SunCertPathBuilder.java:129)
	at java.base/java.security.cert.CertPathBuilder.build(CertPathBuilder.java:297)
	at java.base/sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:383)
	at java.base/sun.security.validator.PKIXValidator.engineValidate(PKIXValidator.java:271)
	at java.base/sun.security.validator.Validator.validate(Validator.java:256)
	at java.base/sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:241)
	at java.base/sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(X509TrustManagerImpl.java:113)
	at io.ktor.network.tls.TLSClientHandshake.handleCertificatesAndKeys(TLSClientHandshake.kt:235)
	at io.ktor.network.tls.TLSClientHandshake.access$handleCertificatesAndKeys(TLSClientHandshake.kt:24)
	at io.ktor.network.tls.TLSClientHandshake$handleCertificatesAndKeys$1.invokeSuspend(TLSClientHandshake.kt)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
	at kotlinx.coroutines.internal.SoftLimitedDispatcher$Worker.run(SoftLimitedDispatcher.kt:125)
	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:99)
	... 4 more
2024-09-21 13:04:59,278 [  18071] SEVERE - #c.i.i.s.i.StartupManagerImpl - JetBrains Rider 2024.2.5  Build #RD-242.22855.90
2024-09-21 13:04:59,278 [  18071] SEVERE - #c.i.i.s.i.StartupManagerImpl - JDK: 21.0.3; VM: OpenJDK 64-Bit Server VM; Vendor: JetBrains s.r.o.
2024-09-21 13:04:59,278 [  18071] SEVERE - #c.i.i.s.i.StartupManagerImpl - OS: Windows 10
2024-09-21 13:04:59,278 [  18071] SEVERE - #c.i.i.s.i.StartupManagerImpl - Plugin to blame: Azure Toolkit for Rider version: 4.2.0
2024-09-21 13:04:59,829 [  18622]   INFO - #PluginsAdvertiser - Plugin is ignored by user, suggestion will not be shown: fr.socolin.application-insights-debug-log-viewer
2024-09-21 13:04:59,829 [  18622]   INFO - #PluginsAdvertiser - Plugin is ignored by user, suggestion will not be shown: fr.socolin.application-insights-debug-log-viewer
2024-09-21 13:04:59,829 [  18622]   INFO - #PluginsAdvertiser - Plugin is ignored by user, suggestion will not be shown: com.intellij.kubernetes
2024-09-21 13:05:00,086 [  18879]   INFO - #c.i.k.a.KubernetesApiProvider - Configuration applied to executor
2024-09-21 13:05:00,089 [  18882]   INFO - #c.i.k.a.KubernetesApiProvider - Configuration refresh finished
```

Any idea what/where is it trying to connect to. or validate with, so I can sort out the relevant certificate for it? Assuming this is something I can even do... could it be the plugin was signed with a different cert, or something?

**Edit to add:** I am not sure if it's related, but I've also discovered that it seems to have clobbered all my Azure Function launch configurations. Previously there used to be at least one 'lightning bolt' Azure function run/debug config entry for each function. They've all gone now, replaced by a series of '?' configs which it says it can't edit. 

Thanks :)

--- a/PluginsAndFeatures/azure-toolkit-for-rider/CHANGELOG.md
+++ b/PluginsAndFeatures/azure-toolkit-for-rider/CHANGELOG.md
@@ -4,6 +4,10 @@
 
 ## [Unreleased]
 
+### Fixed
+
+- Error: unable to find valid certification path to requested target ([#788](https://github.com/JetBrains/azure-tools-for-intellij/issues/788))
+
 ## [4.2.0] - 2024-09-19
 
 ### Fixed

--- a/PluginsAndFeatures/azure-toolkit-for-rider/azure-intellij-plugin-appservice-dotnet/src/main/kotlin/com/microsoft/azure/toolkit/intellij/legacy/function/coreTools/FunctionCoreToolsReleaseFeedService.kt
+++ b/PluginsAndFeatures/azure-toolkit-for-rider/azure-intellij-plugin-appservice-dotnet/src/main/kotlin/com/microsoft/azure/toolkit/intellij/legacy/function/coreTools/FunctionCoreToolsReleaseFeedService.kt
@@ -2,17 +2,24 @@
  * Copyright 2018-2023 JetBrains s.r.o. and contributors. Use of this source code is governed by the MIT license.
  */
 
+@file:OptIn(ExperimentalSerializationApi::class)
+
 package com.microsoft.azure.toolkit.intellij.legacy.function.coreTools
 
-import com.google.gson.GsonBuilder
-import com.google.gson.annotations.SerializedName
 import com.intellij.openapi.Disposable
 import com.intellij.openapi.components.Service
 import com.intellij.openapi.components.service
+import com.intellij.util.net.ssl.CertificateManager
 import io.ktor.client.*
 import io.ktor.client.call.*
 import io.ktor.client.engine.cio.*
+import io.ktor.client.plugins.contentnegotiation.*
 import io.ktor.client.request.*
+import io.ktor.serialization.kotlinx.json.*
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.json.Json
 
 @Service(Service.Level.APP)
 class FunctionCoreToolsReleaseFeedService : Disposable {
@@ -20,50 +27,66 @@ class FunctionCoreToolsReleaseFeedService : Disposable {
         fun getInstance(): FunctionCoreToolsReleaseFeedService = service()
     }
 
-    private val client = HttpClient(CIO)
+    private val client = HttpClient(CIO) {
+        engine {
+            https {
+                trustManager = CertificateManager.getInstance().trustManager
+            }
+        }
+        install(ContentNegotiation) {
+            json(Json {
+                explicitNulls = false
+                ignoreUnknownKeys = true
+            })
+        }
+    }
 
     suspend fun getReleaseFeed(feedUrl: String): ReleaseFeed {
-        val body: String = client.get(feedUrl).body()
-        return GsonBuilder().create().fromJson(body, ReleaseFeed::class.java)
+        val response = client.get(feedUrl)
+        return response.body<ReleaseFeed>()
     }
 
     override fun dispose() = client.close()
 }
 
+@Serializable
 data class ReleaseFeed(
-    @SerializedName("tags")
+    @SerialName("tags")
     val tags: Map<String, Tag>,
-    @SerializedName("releases")
+    @SerialName("releases")
     val releases: Map<String, Release>
 )
 
+@Serializable
 data class Tag(
-    @SerializedName("release")
+    @SerialName("release")
     val release: String?,
-    @SerializedName("releaseQuality")
+    @SerialName("releaseQuality")
     val releaseQuality: String?,
-    @SerializedName("hidden")
+    @SerialName("hidden")
     val hidden: Boolean
 )
 
+@Serializable
 data class Release(
-    @SerializedName("templates")
+    @SerialName("templates")
     val templates: String?,
-    @SerializedName("coreTools")
+    @SerialName("coreTools")
     val coreTools: List<ReleaseCoreTool>
 )
 
+@Serializable
 data class ReleaseCoreTool(
-    @SerializedName("OS")
+    @SerialName("OS")
     val os: String?,
-    @SerializedName("Architecture")
+    @SerialName("Architecture")
     val architecture: String?,
-    @SerializedName("downloadLink")
+    @SerialName("downloadLink")
     val downloadLink: String?,
-    @SerializedName("sha2")
+    @SerialName("sha2")
     val sha2: String?,
-    @SerializedName("size")
+    @SerialName("size")
     val size: String?,
-    @SerializedName("default")
+    @SerialName("default")
     val default: Boolean
 )

--- a/PluginsAndFeatures/azure-toolkit-for-rider/azure-intellij-plugin-cloud-shell/build.gradle.kts
+++ b/PluginsAndFeatures/azure-toolkit-for-rider/azure-intellij-plugin-cloud-shell/build.gradle.kts
@@ -15,7 +15,6 @@ repositories {
 }
 
 val platformVersion: String by extra
-val azureToolkitVersion: String by extra
 
 dependencies {
     intellijPlatform {

--- a/PluginsAndFeatures/azure-toolkit-for-rider/gradle.properties
+++ b/PluginsAndFeatures/azure-toolkit-for-rider/gradle.properties
@@ -4,7 +4,7 @@ pluginGroup = com.jetbrains
 pluginName = azure-toolkit-for-rider
 pluginRepositoryUrl = https://github.com/JetBrains/azure-tools-for-intellij
 # SemVer format -> https://semver.org
-pluginVersion = 4.2.0
+pluginVersion = 4.2.1
 
 # Supported build number ranges and IntelliJ Platform versions -> https://plugins.jetbrains.com/docs/intellij/build-number-ranges.html
 pluginSinceBuild = 242
@@ -15,12 +15,12 @@ pluginUntilBuild = 242.*
 #   Release:  2020.2
 #   EAP:      2020.3-EAP2-SNAPSHOT
 #   Nightly:  2020.3-SNAPSHOT
-platformVersion = 2024.2
+platformVersion = 2024.2.5
 
 dotnetBuildConfiguration=Release
 
 # Gradle Releases -> https://github.com/gradle/gradle/releases
-gradleVersion = 8.10
+gradleVersion = 8.10.2
 
 # Opt-out flag for bundling Kotlin standard library -> https://jb.gg/intellij-platform-kotlin-stdlib
 kotlin.stdlib.default.dependency = false

--- a/PluginsAndFeatures/azure-toolkit-for-rider/gradle/wrapper/gradle-wrapper.properties
+++ b/PluginsAndFeatures/azure-toolkit-for-rider/gradle/wrapper/gradle-wrapper.properties
@@ -1,6 +1,6 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-8.10-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.10.2-bin.zip
 networkTimeout=10000
 validateDistributionUrl=true
 zipStoreBase=GRADLE_USER_HOME

--- a/PluginsAndFeatures/azure-toolkit-for-rider/src/dotnet/Directory.Build.props
+++ b/PluginsAndFeatures/azure-toolkit-for-rider/src/dotnet/Directory.Build.props
@@ -8,7 +8,7 @@
     <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
     <ResolveAssemblyWarnOrErrorOnTargetArchitectureMismatch>None</ResolveAssemblyWarnOrErrorOnTargetArchitectureMismatch>
 
-    <RiderSDKVersion>2024.2.0</RiderSDKVersion>
+    <RiderSDKVersion>2024.2.2</RiderSDKVersion>
   </PropertyGroup>
 
   <PropertyGroup Condition="'$(Configuration)'=='Debug'">

